signac/core/syncedattrdict.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from collections.abc import Mapping
+from collections.abc import MutableMapping
+
+from .collection_api import SyncedCollection
+
+
+class SyncedAttrDict(SyncedCollection, MutableMapping):
+    """Implements the dict data structures"""
+
+    base_type = 'mapping'  # type: ignore
+
+    _PROTECTED_KEYS = ('_data', '_suspend_sync_', '_load', '_sync', '_parent')
+
+    VALID_KEY_TYPES = (str, int, bool, type(None))
+
+    def __init__(self, data=None, **kwargs):
+        super().__init__(**kwargs)
+        if data is None:
+            self._data = {}
+        else:
+            with self._suspend_sync():
+                self._data = {
+                    self._validate_key(key): self.from_base(data=value, parent=self)
+                    for key, value in data.items()
+                }
+            self.sync()
+
+    def to_base(self):
+        """Converts the SyncedDict object to Dictionary.
+
+        Returns:
+        --------
+        converted: dict
+            Dictionary containing the converted synced dict object."""
+        converted = {}
+        for key, value in self._data.items():
+            if isinstance(value, SyncedCollection):
+                converted[key] = value.to_base()
+            else:
+                converted[key] = value
+        return converted
+
+    @classmethod
+    def is_base_type(cls, data):
+        """Checks whether the data is of base type of SyncedDict
+        Parameters
+        ----------
+        data: any
+            Data to be checked.
+
+        Returns
+        -------
+        bool
+        """
+        if isinstance(data, Mapping):
+            return True
+        return False
+
+    def _dfs_update(self, data=None):
+        """Updates the SyncedDict instance with data by using dfs."""
+        if data is None:
+            data = {}
+        if isinstance(data, Mapping):
+            with self._suspend_sync():
+                for key in data:
+                    if key in self._data:
+                        if data[key] == self._data[key]:
+                            continue
+                        if isinstance(self._data[key], SyncedCollection):
+                            try:
+                                self._data[key]._dfs_update(data[key])
+                                continue
+                            except ValueError:
+                                pass
+                    self[key] = data[key]
```

Comment:
In the list implementation, you directly call setitem on `_data`, can we do that here too since we're suspending synchronization?